server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -379,64 +376,88 @@ static boolean executeBulkItemRequest(
                 request.getAutoGeneratedTimestamp(),
                 request.isRetry()
             );
-
+            if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
+                return handleMappingUpdateRequired(
+                    context,
+                    mappingUpdater,
+                    waitForMappingUpdate,
+                    itemDoneListener,
+                    result,
+                    primary,
+                    version,
+                    updateResult
+                );
+            }
         }
-        if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
 
-            try {
-                Optional<CompressedXContent> mergedSource = Optional.ofNullable(
-                    primary.mapperService()
-                        .merge(
-                            MapperService.SINGLE_MAPPING_NAME,
-                            new CompressedXContent(result.getRequiredMappingUpdate()),
-                            MapperService.MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT
-                        )
-                ).map(DocumentMapper::mappingSource);
-                Optional<CompressedXContent> previousSource = Optional.ofNullable(primary.mapperService().documentMapper())
-                    .map(DocumentMapper::mappingSource);
+        onComplete(result, context, updateResult);
+        return true;
+    }
 
-                if (mergedSource.equals(previousSource)) {
-                    context.resetForNoopMappingUpdateRetry(primary.mapperService().mappingVersion());
-                    return true;
-                }
-            } catch (Exception e) {
-                logger.info(() -> format("%s mapping update rejected by primary", primary.shardId()), e);
-                assert result.getId() != null;
-                onComplete(exceptionToResult(e, primary, isDelete, version, result.getId()), context, updateResult);
+    private static boolean handleMappingUpdateRequired(
```

Comment:
Just extracted this into its own method now that it got so big and removed the `Optional` use which was confusing and unnecessary. Sorry for mixing fix + refactoring a little here, hope it's not too confusing. Otherwise I can make this into a 2 parter.

Reply:
given how delicate things are, I think it would be great to split this in two: refactoring  + actual change, as a safety measure. I could definitely miss things otherwise while reviewing this as-is.

Reply:
thanks!